mailbox/store/src/main/java/org/apache/james/mailbox/store/mail/model/impl/MessageParser.java

```diff
@@ -67,35 +64,44 @@
         MessageWriter messageWriter = new DefaultMessageWriter();
         for (Entity entity : multipart.getBodyParts()) {
             if (isAttachment(entity)) {
-                Optional<String> contentType = contentType(entity.getHeader().getField(CONTENT_TYPE));
-                attachments.add(createAttachment(messageWriter, entity.getBody(), contentType.or(DEFAULT_CONTENT_TYPE)));
+                Optional<ContentTypeField> contentTypeField = contentTypeField(entity.getHeader().getField(CONTENT_TYPE));
+                Optional<String> contentType = contentType(contentTypeField);
+                Optional<String> name = name(contentTypeField);
+                attachments.add(createAttachment(messageWriter, entity.getBody(), contentType.or(DEFAULT_CONTENT_TYPE), name));
             }
         }
         return attachments.build();
     }
 
-    private Optional<String> contentType(Field contentType) {
-        if (contentType == null) {
-            return Optional.absent();
+    private Optional<String> contentType(Optional<ContentTypeField> contentTypeField) {
+        if (contentTypeField.isPresent()) {
+            return Optional.fromNullable(contentTypeField.get().getMimeType());
         }
-        String body = contentType.getBody();
-        if (Strings.isNullOrEmpty(body)) {
-            return Optional.absent();
+        return Optional.absent();
+    }
+
+    private Optional<String> name(Optional<ContentTypeField> contentType) {
+        if (contentType.isPresent()) {
+            return Optional.fromNullable(contentType.get().getParameter("name"));
         }
-        if (body.contains(HEADER_SEPARATOR)) {
-            return Optional.of(Iterables.get(Splitter.on(HEADER_SEPARATOR).split(body), 0));
+        return Optional.absent();
+    }
+
+    private Optional<ContentTypeField> contentTypeField(Field contentType) {
+        if (contentType == null || !(contentType instanceof ContentTypeField)) {
+            return Optional.absent();
         }
-        return Optional.of(body);
+        return Optional.of((ContentTypeField) contentType);
     }
 
     private boolean isAttachment(Entity part) {
         return ContentDispositionField.DISPOSITION_TYPE_ATTACHMENT.equalsIgnoreCase(part.getDispositionType());
     }
 
-    private Attachment createAttachment(MessageWriter messageWriter, Body body, String contentType) throws IOException {
+    private Attachment createAttachment(MessageWriter messageWriter, Body body, String contentType, Optional<String> name) throws IOException {
```

Comment:
this method should not exist, you should use the builder to create the attachment. The only interesting part is the one that generates a stream, you should keep that in a method.